sdk/servicebus/Azure.Messaging.ServiceBus/samples/Sample15_MockingClientTypes.md

```diff
@@ -0,0 +1,1262 @@
+
+ServiceBusReceivedMessage receivedMessage1 = await receiver.ReceiveMessageAsync(TimeSpan.FromSeconds(1), CancellationToken.None);
+await receiver.CompleteMessageAsync(receivedMessage1);
+
+ServiceBusReceivedMessage receivedMessage2 = await receiver.ReceiveMessageAsync(TimeSpan.FromSeconds(1), CancellationToken.None);
+await receiver.CompleteMessageAsync(receivedMessage2);
+
+// For illustrative purposes, verify that the number of times a message was received is the same number of times
+// a message was completed.
+
+mockReceiver
+    .Verify(receiver => receiver.CompleteMessageAsync(
+        It.IsAny<ServiceBusReceivedMessage>(),
+        It.IsAny<CancellationToken>()),
+        Times.Exactly(numCallsReceiveMessage));
+```
+
+```C# Snippet:ServiceBus_MockDefer
+// The first section sets up a mock ServiceBusReceiver.
+
+Mock<ServiceBusClient> mockClient = new();
+Mock<ServiceBusReceiver> mockReceiver = new();
+
+mockClient
+    .Setup(client => client.CreateReceiver(It.IsAny<string>()))
+    .Returns(mockReceiver.Object);
+
+ServiceBusClient client = mockClient.Object;
+
+// This creates a list of messages to return from the ServiceBusReceiver mock. See the ServiceBusModelFactory
+// for a complete set of properties that can be populated using the ServiceBusModelFactory.ServiceBusReceivedMessage method.
+
+List<ServiceBusReceivedMessage> messagesToReturn = new();
+int numMessages = 3;
+
+for (int i = 0; i < numMessages; i++)
+{
+    string body = $"message-{i}";
+
+    // This mocks a ServiceBusReceivedMessage instance using the model factory. Different arguments can mock different
+    // potential outputs from the broker.
+
+    ServiceBusReceivedMessage messageToReturn = ServiceBusModelFactory.ServiceBusReceivedMessage(
+        body: new BinaryData(body),
+        messageId: $"id-{i}",
+        sequenceNumber: i, // this needs to be populated in order to defer each message
+        partitionKey: "illustrative-partitionKey",
+        correlationId: "illustrative-correlationId",
+        contentType: "illustrative-contentType",
+        replyTo: "illustrative-replyTo"
+        // ...
+        );
+    messagesToReturn.Add(messageToReturn);
+}
+
+// This sets up the mock ServiceBusReceiver to return a different message from the messagesToReturn
+// list each time the method is called until there are no more messages.
+
+int numCallsReceiveMessage = 0;
+mockReceiver
+    .Setup(receiver => receiver.ReceiveMessageAsync(
+        It.IsAny<TimeSpan>(),
+        It.IsAny<CancellationToken>()))
+    .ReturnsAsync(() =>
+    {
+        ServiceBusReceivedMessage m = messagesToReturn.ElementAtOrDefault(numCallsReceiveMessage);
+        numCallsReceiveMessage++;
+        return m;
+    });
+
+// This sets up the mock ServiceBusReceiver to keep track of the messages being deferred using each message's
+// sequence number. It creates a new ServiceBusReceivedMessage using the model factory in order to update
```

Comment:
I added it since the application properties are read-only and wanted to demonstrate how to mock updating the properties using DeferMessageAsync. There may be a better way to do that but this was what seemed to work/ make the most sense to me.